File: app/src/main/java/com/waz/zclient/pages/main/connect/SendConnectRequestFragment.java
/**
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.pages.main.connect;

import android.graphics.Color;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.widget.TextView;
import com.waz.api.IConversation;
import com.waz.api.User;
import com.waz.model.UserId;
import com.waz.zclient.R;
import com.waz.zclient.common.controllers.UserAccountsController;
import com.waz.zclient.common.views.UserDetailsView;
import com.waz.zclient.controllers.accentcolor.AccentColorObserver;
import com.waz.zclient.conversation.ConversationController;
import com.waz.zclient.core.stores.connect.ConnectStoreObserver;
import com.waz.zclient.core.stores.connect.IConnectStore;
import com.waz.zclient.pages.BaseFragment;
import com.waz.zclient.pages.main.participants.ProfileAnimation;
import com.waz.zclient.pages.main.participants.dialog.DialogLaunchMode;
import com.waz.zclient.ui.theme.ThemeUtils;
import com.waz.zclient.ui.utils.KeyboardUtils;
import com.waz.zclient.ui.views.ZetaButton;
import com.waz.zclient.utils.ContextUtils;
import com.waz.zclient.utils.ViewUtils;
import com.waz.zclient.views.images.ImageAssetImageView;
import com.waz.zclient.views.menus.FooterMenu;
import com.waz.zclient.views.menus.FooterMenuCallback;

public class SendConnectRequestFragment extends BaseFragment<SendConnectRequestFragment.Container> implements ConnectStoreObserver,
                                                                                                              AccentColorObserver {
    public static final String TAG = SendConnectRequestFragment.class.getName();
    public static final String ARGUMENT_USER_ID = "ARGUMENT_USER_ID";
    public static final String ARGUMENT_USER_REQUESTER = "ARGUMENT_USER_REQUESTER";


    private String userId;
    private IConnectStore.UserRequester userRequester;

    // Flag true if layout has been set
    private TextView displayNameTextView;
    private UserDetailsView userDetailsView;
    private ZetaButton connectButton;
    private FooterMenu footerMenu;
    private ImageAssetImageView imageAssetImageViewProfile;

    public static SendConnectRequestFragment newInstance(String userId, IConnectStore.UserRequester userRequester) {
        SendConnectRequestFragment newFragment = new SendConnectRequestFragment();

        Bundle args = new Bundle();
        args.putString(ARGUMENT_USER_ID, userId);
        args.putString(ARGUMENT_USER_REQUESTER, userRequester.toString());
        newFragment.setArguments(args);

        return newFragment;
    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  Lifecycle
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public Animation onCreateAnimation(int transit, boolean enter, int nextAnim) {
        Animation animation = super.onCreateAnimation(transit, enter, nextAnim);

        if (getControllerFactory().getConversationScreenController().getPopoverLaunchMode() != DialogLaunchMode.AVATAR &&
            getControllerFactory().getConversationScreenController().getPopoverLaunchMode() != DialogLaunchMode.COMMON_USER) {
            int centerX = ContextUtils.getOrientationIndependentDisplayWidth(getActivity()) / 2;
            int centerY = ContextUtils.getOrientationIndependentDisplayHeight(getActivity()) / 2;
            int duration;
            int delay = 0;

            if (nextAnim != 0) {
                if (enter) {
                    duration = getResources().getInteger(R.integer.open_profile__animation_duration);
                    delay = getResources().getInteger(R.integer.open_profile__delay);
                } else {
                    duration = getResources().getInteger(R.integer.close_profile__animation_duration);
                }

                animation = new ProfileAnimation(enter, duration, delay, centerX, centerY);
            }
        }
        return animation;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, final ViewGroup viewContainer, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_send_connect_request, viewContainer, false);

        if (savedInstanceState == null) {
            userId = getArguments().getString(ARGUMENT_USER_ID);
            userRequester = IConnectStore.UserRequester.valueOf(getArguments().getString(ARGUMENT_USER_REQUESTER));
        } else {
            userId = savedInstanceState.getString(ARGUMENT_USER_ID);
            userRequester = IConnectStore.UserRequester.valueOf(savedInstanceState.getString(ARGUMENT_USER_REQUESTER));
        }
        Toolbar toolbar = ViewUtils.getView(rootView, R.id.t__send_connect__toolbar);
        displayNameTextView = ViewUtils.getView(rootView, R.id.tv__send_connect__toolbar__title);
        userDetailsView = ViewUtils.getView(rootView, R.id.udv__send_connect__user_details);
        connectButton = ViewUtils.getView(rootView, R.id.zb__send_connect_request__connect_button);
        footerMenu = ViewUtils.getView(rootView, R.id.fm__footer);
        imageAssetImageViewProfile = ViewUtils.getView(rootView, R.id.iaiv__send_connect);
        imageAssetImageViewProfile.setDisplayType(ImageAssetImageView.DisplayType.CIRCLE);
        imageAssetImageViewProfile.setSaturation(0);

        View backgroundContainer = ViewUtils.getView(rootView, R.id.fl__send_connect_request__background_container);
        backgroundContainer.setClickable(true);
        if (userRequester == IConnectStore.UserRequester.PARTICIPANTS &&
            getControllerFactory().getConversationScreenController().getPopoverLaunchMode() != DialogLaunchMode.AVATAR &&
            getControllerFactory().getConversationScreenController().getPopoverLaunchMode() != DialogLaunchMode.COMMON_USER) {
            backgroundContainer.setBackgroundColor(Color.TRANSPARENT);
        }
        connectButton.setText(getResources().getString(R.string.send_connect_request__connect_button__text));

        if (ThemeUtils.isDarkTheme(getContext())) {
            toolbar.setNavigationIcon(R.drawable.action_back_light);
        } else {
            toolbar.setNavigationIcon(R.drawable.action_back_dark);
        }
        toolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                getContainer().dismissUserProfile();
            }
        });

        footerMenu.setVisibility(View.GONE);

        return rootView;
    }

    @Override
    public void onStart() {
        super.onStart();

        getControllerFactory().getAccentColorController().addAccentColorObserver(this);

        getStoreFactory().connectStore().addConnectRequestObserver(this);
        getStoreFactory().connectStore().loadUser(userId, userRequester);

    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putString(ARGUMENT_USER_ID, userId);
        if (userRequester != null) {
            outState.putString(ARGUMENT_USER_REQUESTER, userRequester.toString());
        }
        super.onSaveInstanceState(outState);
    }


    @Override
    public void onStop() {
        KeyboardUtils.hideKeyboard(getActivity());

        getStoreFactory().connectStore().removeConnectRequestObserver(this);
        getControllerFactory().getAccentColorController().removeAccentColorObserver(this);

        super.onStop();
    }

    @Override
    public void onDestroyView() {
        imageAssetImageViewProfile = null;
        super.onDestroyView();
    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  IConnectStoreObserver
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onConnectUserUpdated(final User user, final IConnectStore.UserRequester userRequester) {
        if (this.userRequester != userRequester) {
            return;
        }

        imageAssetImageViewProfile.connectImageAsset(user.getPicture());

        displayNameTextView.setText(user.getName());
        userDetailsView.setUser(user);

        connectButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (user == null ||
                    getStoreFactory() == null ||
                    getStoreFactory().isTornDown()) {
                    return;
                }
                User me = getStoreFactory().profileStore().getSelfUser();
                String myName = me != null ? me.getName() : "";
                String otherName = user.getName();
                String message = getString(R.string.connect__message, otherName, myName);
                IConversation conversation = getStoreFactory().connectStore().connectToNewUser(user, message);
                if (conversation != null) {
                    KeyboardUtils.hideKeyboard(getActivity());
                    getContainer().onConnectRequestWasSentToUser();
                }
            }
        });

        final Boolean permissionToRemove = inject(UserAccountsController.class).hasRemoveConversationMemberPermission(
            inject(ConversationController.class).getCurrentConvId()
        );

        if (userRequester == IConnectStore.UserRequester.PARTICIPANTS && permissionToRemove) {
            footerMenu.setRightActionText(getString(R.string.glyph__minus));
        }

        footerMenu.setCallback(new FooterMenuCallback() {
            @Override
            public void onLeftActionClicked() {
                showConnectButtonInsteadOfFooterMenu();
            }

            @Override
            public void onRightActionClicked() {
                if (userRequester == IConnectStore.UserRequester.PARTICIPANTS && permissionToRemove) {
                    getContainer().showRemoveConfirmation(new UserId(user.getId()));
                }
            }
        });

        if (userRequester == IConnectStore.UserRequester.PARTICIPANTS) {
            footerMenu.setVisibility(View.VISIBLE);
            connectButton.setVisibility(View.GONE);
        } else {
            footerMenu.setVisibility(View.GONE);
            connectButton.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onInviteRequestSent(IConversation conversation) {

    }

    //////////////////////////////////////////////////////////////////////////////////////////
    //
    //  AccentColorObserver
    //
    //////////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onAccentColorHasChanged(Object sender, int color) {
        connectButton.setAccentColor(color);
    }

    private void showConnectButtonInsteadOfFooterMenu() {
        if (connectButton.getVisibility() == View.VISIBLE) {
            return;
        }
        footerMenu.setVisibility(View.GONE);
        connectButton.setAlpha(0);
        connectButton.setVisibility(View.VISIBLE);
        ViewUtils.fadeInView(connectButton,
                             getResources().getInteger(R.integer.framework_animation_duration_long));
    }

    public interface Container extends UserProfileContainer {
        void onConnectRequestWasSentToUser();

        void showRemoveConfirmation(UserId userId);
    }
}

File: app/src/main/scala/com/waz/zclient/connect/SendConnectRequestFragment.scala
/**
 * Wire
 * Copyright (C) 2018 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.waz.zclient.connect

import android.content.Context
import android.graphics.Color
import android.os.Bundle
import android.view.animation.Animation
import android.view.{LayoutInflater, View, ViewGroup}
import android.widget.ImageView
import com.waz.ZLog.ImplicitTag._
import com.waz.model.UserId
import com.waz.service.ZMessaging
import com.waz.threading.Threading
import com.waz.utils.events.{ClockSignal, Signal}
import com.waz.utils.returning
import com.waz.zclient.common.controllers.global.KeyboardController
import com.waz.zclient.common.controllers.global.AccentColorController
import com.waz.zclient.common.controllers.{ThemeController, UserAccountsController}
import com.waz.zclient.common.views.ImageAssetDrawable
import com.waz.zclient.common.views.ImageAssetDrawable.{RequestBuilder, ScaleType}
import com.waz.zclient.common.views.ImageController.WireImage
import com.waz.zclient.connect.PendingConnectRequestFragment.ArgUserRequester
import com.waz.zclient.conversation.ConversationController
import com.waz.zclient.core.stores.connect.IConnectStore
import com.waz.zclient.messages.UsersController
import com.waz.zclient.pages.BaseFragment
import com.waz.zclient.pages.main.connect.UserProfileContainer
import com.waz.zclient.pages.main.participants.ProfileAnimation
import com.waz.zclient.paintcode.GuestIcon
import com.waz.zclient.ui.text.TypefaceTextView
import com.waz.zclient.ui.views.ZetaButton
import com.waz.zclient.utils.ContextUtils._
import com.waz.zclient.utils.{GuestUtils, StringUtils, ViewUtils}
import com.waz.zclient.views.menus.{FooterMenu, FooterMenuCallback}
import com.waz.zclient.{FragmentHelper, R}
import org.threeten.bp.Instant
import scala.concurrent.duration._

class SendConnectRequestFragment extends BaseFragment[SendConnectRequestFragment.Container]
  with FragmentHelper {

  import SendConnectRequestFragment._
  import Threading.Implicits.Ui

  implicit def context: Context = getActivity

  private lazy val userToConnectId = UserId(getArguments.getString(ArgumentUserId))
  private lazy val userRequester = IConnectStore.UserRequester.valueOf(getArguments.getString(ArgumentUserRequester))

  private lazy val usersController = inject[UsersController]
  private lazy val userAccountsController = inject[UserAccountsController]
  private lazy val conversationController = inject[ConversationController]
  private lazy val keyboardController = inject[KeyboardController]
  private lazy val accentColorController = inject[AccentColorController]
  private lazy val zms = inject[Signal[ZMessaging]]
  private lazy val themeController = inject[ThemeController]

  private lazy val user = usersController.user(userToConnectId)

  private lazy val removeConvMemberFeatureEnabled = for {
    convId <- conversationController.currentConvId
    permission <- userAccountsController.hasRemoveConversationMemberPermission(convId)
  } yield permission && userRequester == IConnectStore.UserRequester.PARTICIPANTS

  private lazy val connectButton = returning(view[ZetaButton](R.id.zb__send_connect_request__connect_button)) { vh =>
    accentColorController.accentColor.onUi { color => vh.foreach(_.setAccentColor(color.getColor)) }
    vh.onClick { _ =>
      usersController.connectToUser(userToConnectId).foreach(_.foreach { _ =>
        keyboardController.hideKeyboardIfVisible()
        getContainer.onConnectRequestWasSentToUser()
      })
    }
  }
  private lazy val footerMenu = returning(view[FooterMenu](R.id.fm__footer)) { vh =>
    Signal(removeConvMemberFeatureEnabled, user.map(_.expiresAt.isDefined)).onUi {
      case (removeMemberEnabled, isWireless) =>

        val leftText = if (isWireless) "" else getString(R.string.send_connect_request__connect_button__text)
        val leftGlyph = if (isWireless) "" else getString(R.string.glyph__plus)
        val rightGlyph = if (removeMemberEnabled) getString(R.string.glyph__minus) else ""

        vh.foreach { footer =>
          footer.setLeftActionLabelText(leftText)
          footer.setLeftActionText(leftGlyph)
          footer.setRightActionText(rightGlyph)

          footer.setCallback(new FooterMenuCallback() {
            override def onLeftActionClicked(): Unit = {
              if (!isWireless)
                showConnectButtonInsteadOfFooterMenu()
            }
            override def onRightActionClicked(): Unit = {
              if (removeMemberEnabled)
                getContainer.showRemoveConfirmation(userToConnectId)
            }
          })
        }
    }
  }
  private lazy val imageViewProfile = view[ImageView](R.id.send_connect)
  private lazy val userNameView = returning(view[TypefaceTextView](R.id.user_name)) { vh =>
    user.map(_.getDisplayName).onUi(t => vh.foreach(_.setText(t)))
  }
  private lazy val userHandleView = returning(view[TypefaceTextView](R.id.user_handle)) { vh =>
    user.map(user => StringUtils.formatHandle(user.handle.map(_.string).getOrElse("")))
      .onUi(t => vh.foreach(_.setText(t)))
  }
  private lazy val guestIndicator = returning(view[View](R.id.guest_indicator)) { indicator =>
    zms.flatMap(z => user.map(_.isGuest(z.teamId))).map {
      case true => View.VISIBLE
      case _ => View.GONE
    }.onUi(indicator.setVisibility(_))
  }
  private lazy val guestIndicatorIcon = view[ImageView](R.id.guest_indicator_icon)
  private lazy val guestIndicatorTimer = returning(view[TypefaceTextView](R.id.guest_indicator_timer)) { text =>
    (for {
      expires <- user.map(_.expiresAt)
      clock <- if (expires.isDefined) ClockSignal(5.minutes) else Signal.const(Instant.EPOCH)
    } yield expires match {
      case Some(expiresAt) => GuestUtils.timeRemainingString(expiresAt, clock)
      case _ => ""
    }).onUi(t => text.foreach(_.setText(t)))
  }

  override def onCreateAnimation(transit: Int, enter: Boolean, nextAnim: Int): Animation = {
    def defaultAnimation = super.onCreateAnimation(transit, enter, nextAnim)
    def isConvRequester = {
      val userRequester = IConnectStore.UserRequester.valueOf(getArguments.getString(ArgUserRequester))
      userRequester == IConnectStore.UserRequester.CONVERSATION
    }

    if (isConvRequester || nextAnim != 0) defaultAnimation
    else {
      val centerX = getOrientationIndependentDisplayWidth(getActivity) / 2
      val centerY = getOrientationIndependentDisplayHeight(getActivity) / 2
      val duration =
        if (enter) getInt(R.integer.open_profile__animation_duration)
        else getInt(R.integer.close_profile__animation_duration)
      val delay =
        if (enter) getInt(R.integer.open_profile__delay)
        else 0

      new ProfileAnimation(enter, duration, delay, centerX, centerY)
    }
  }

  override def onCreateView(inflater: LayoutInflater, viewContainer: ViewGroup, savedInstanceState: Bundle): View =
    inflater.inflate(R.layout.fragment_send_connect_request, viewContainer, false)

  override def onViewCreated(view: View, savedInstanceState: Bundle): Unit = {
    userNameView
    userHandleView
    guestIndicator
    guestIndicatorTimer

    val color = if (themeController.isDarkTheme) R.color.wire__text_color_primary_dark_selector else R.color.wire__text_color_primary_light_selector
    guestIndicatorIcon.foreach(_.setImageDrawable(GuestIcon(color)))

    val assetDrawable = new ImageAssetDrawable(
      user.map(_.picture).collect { case Some(p) => WireImage(p) },
      scaleType = ScaleType.CenterInside,
      request = RequestBuilder.Round
    )
    imageViewProfile.foreach(_.setImageDrawable(assetDrawable))

    val backgroundContainer = findById[View](R.id.background_container)
    backgroundContainer.setClickable(true)

    if (userRequester == IConnectStore.UserRequester.PARTICIPANTS) {
      backgroundContainer.setBackgroundColor(Color.TRANSPARENT)
      footerMenu.setVisibility(View.VISIBLE)
      connectButton.setVisibility(View.GONE)
    } else {
      footerMenu.setVisibility(View.GONE)
      connectButton.setVisibility(View.VISIBLE)
    }

    footerMenu.foreach(_.setCallback(new FooterMenuCallback {
      override def onLeftActionClicked(): Unit = showConnectButtonInsteadOfFooterMenu()
      override def onRightActionClicked(): Unit = removeConvMemberFeatureEnabled.head foreach {
        case true => getContainer.showRemoveConfirmation(userToConnectId)
        case _ =>
      }
    }))
  }

  override def onStop(): Unit = {
    keyboardController.hideKeyboardIfVisible()
    super.onStop()
  }

  private def showConnectButtonInsteadOfFooterMenu(): Unit = {
    if (connectButton.getVisibility != View.VISIBLE) {
      footerMenu.foreach(_.setVisibility(View.GONE))
      connectButton.foreach { btn =>
        btn.setAlpha(0)
        btn.setVisibility(View.VISIBLE)
        ViewUtils.fadeInView(btn, getInt(R.integer.framework_animation_duration_long))
      }
    }
  }
}

object SendConnectRequestFragment {
  val Tag: String = classOf[SendConnectRequestFragment].getName
  val ArgumentUserId = "ARGUMENT_USER_ID"
  val ArgumentUserRequester = "ARGUMENT_USER_REQUESTER"

  def newInstance(userId: String, userRequester: IConnectStore.UserRequester): SendConnectRequestFragment =
    returning(new SendConnectRequestFragment)(fragment =>
      fragment.setArguments(returning(new Bundle) { args =>
        args.putString(ArgumentUserId, userId)
        args.putString(ArgumentUserRequester, userRequester.toString)
      })
    )

  trait Container extends UserProfileContainer {
    def onConnectRequestWasSentToUser(): Unit
    override def showRemoveConfirmation(userId: UserId): Unit
  }

}
